src/backends/simple.ts

```diff
@@ -0,0 +1,442 @@
+                timestamp: targetPTS,
+                any: false,
+            });
+            this._createDecoder();
+            this.packet = null;
+            this.frames = [];
+        }
+
+        // the decoder has been previously flushed while retrieving frames at the end of the stream and has thus been
+        // destroyed. See if the requested targetPTS is part of the last few frames we decoded. If so, return it.
+        if (!this.decoder) {
+            LOG && console.log('no decoder');
+            if ((this.filteredFrames as any).length > 0) {
+                const closestFrame = (this.filteredFrames as any).find(f => f.pts <= targetPTS);
+                // we should probably check the delta between the targetPTS and the closestFrame. If it's too big, we
+                // should return null or something.
+                LOG && console.log('returning closest frame with pts', closestFrame.pts);
+                LOG && console.log('read', packetReadCount, 'packets');
+                this.previousTargetPTS = targetPTS;
+                return closestFrame;
+            }
+            throw Error('Unexpected condition: no decoder and no frames');
+        }
+
+        // Read packets until we have a frame which is closest to targetPTS
+        let filteredFrames = null;
+        let closestFramePTS = -1;
+        let outputFrame = null;
+
+        // This is the first time we're decoding frames. Get the first packet and decode it.
+        if (!this.packet && this.frames.length === 0) {
+            // LOG && console.log('getting initial packet and frames');
+            ({ packet: this.packet, frames: this.frames } = await this._getNextPacketAndDecodeFrames());
+            packetReadCount++;
+        }
+        // LOG && console.log('packet', !!this.packet, 'frames', this.frames.length, 'pts', this.packet.pts);
+
+        while ((this.packet || this.frames.length !== 0) && closestFramePTS < targetPTS && packetReadCount < MAX_PACKET_READS) {
+            LOG && console.log('packet si:', this.packet?.stream_index, 'pts:', this.packet?.pts, 'frames:', this.frames?.length);
+            LOG && console.log('frames', this.frames?.length, 'frames.pts:', this.frames?.map(f => f.pts), '-> target.pts:', targetPTS);
+
+            // packet contains frames
+            if (this.frames.length !== 0) {
+                // filter the frames
+                const filteredResult = await this.filterer.filter([{ name: 'in0:v', frames: this.frames }]);
+                filteredFrames = filteredResult.flatMap(r => r.frames);
+                LOG && console.log('filteredFrames', filteredFrames.length, 'filteredFrames.pts:', filteredFrames.map(f => f.pts), '-> target.pts:', targetPTS);
+                this.filteredFrames = filteredFrames as beamcoder.DecodedFrames;
+
+                // get closest frame
+                const closestFrame = filteredFrames.reverse().find(f => f.pts <= targetPTS);
```

Comment:
I'm not sure I understand the concept of `closestFrame` here. Looks like it is actually the closest frame before the target?

I would love a comment explaining how it's used

Reply:
I added some explanation. Let me know if it's enough